### PR TITLE
lists:subtract (operator --) trapping

### PR DESCRIPTION
**Pull request by kvakvs**
* Before transforming algorithm into a trappable FSM improved `lists:subtract` (`--`) performance by ~11% on random inputs by reversing the algorithm (instead of eliminating elements in A for each element in B, I now shrink B by moving last element in place of eliminated element). On sorted inputs the performance is actually 5-10% lower.
* Algorithm is now a 5-state FSM: 1) Calculate length A; 2) Calculate length B; 3) Copy B to temp array; 4) Subtract B from A, shrinking B as elements in B are consumed; 5) Copy result array to result on process heap.
* All FSM steps are re-entrable and interruptible ( and use some reasonable numbers to limit work (300-600 elements for copying or comparing until it will BIF_TRAP).
* Trap re-enters via `lists:subtract/2` with first arg=`'bif_return_trap'` and second arg `{A, B, MagicBinary}`
* Algorithm for length is trappable and pure and can be reused elsewhere.
* Algorithms for copy list to array and copy array to list are also pure and can be reused elsewhere.
* Should be fairly easily back-portable: Only adds a large block of code to `erl_bif_lists.c` and constructor call to `erl_init`.

### CODE REVIEW

File: erts/emulator/beam/erl_bif_lists.c [hunk 1]
```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
...
+        iter_a = CDR(ptr_a);
+        reds--;
+    }
+
+    if (reds) {
+        /* Reds not 0, work seems to be completed */
+        return am_true;
+    }
+    s->iter_a = iter_a; /* save iterator */
+    return am_false;
```
Comment (jhogberg) [inline erts/emulator/beam/erl_bif_lists.c]: Doesn't this function return `int`?

File: erts/emulator/beam/erl_bif_lists.c [hunk 2]
```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
...
+                BIF_RET(orig_A);
+            }
+            context->fsm_state = SUBTRACT_STEP_BUILD_RESULT;
+            erts_trappable_copy_array_to_list_setup(p, context->result_array,
+                                                    context->s.scan.result_size,
+                                                    &context->s.build_res);
+            /* FALL THROUGH */
+        }
+
+        case SUBTRACT_STEP_BUILD_RESULT: {
```
Comment (jhogberg) [inline erts/emulator/beam/erl_bif_lists.c]: I think you could skip this step if you build the list destructively during SUBTRACT_STEP_SCAN instead. You've already disabled the GC, so it should be safe as long as you never leak the unfinished result.

File: erts/emulator/beam/external.c [hunk 3]
```diff
@@ -1947,13 +1947,14 @@ static Eterm erts_term_to_binary_int(Process* p, Eterm Term, int level, Uint fla
 	    context_b = erts_create_magic_binary(sizeof(TTBContext),    \
                                                  ttb_context_destructor);   \
 	    context =  ERTS_MAGIC_BIN_DATA(context_b);			\
-	    sys_memcpy(context,&c_buff,sizeof(TTBContext));			\
+	    sys_memcpy(context,&c_buff,sizeof(TTBContext));		\
```
Comment (jhogberg) [inline erts/emulator/beam/external.c]: The changes to this file don't look related to the rest of this commit, can you break them out?

File: erts/emulator/beam/erl_bif_lists.c [hunk 4]
```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
...
+ * Returns: am_false - if you need to call it again.
+ */
+static Eterm
+erts_trappable_list_length(ErtsLengthContext *context) {
+    Eterm iter = context->iter;
+    const static Uint WORK_BEFORE_TRAPPING = IF_DEBUG(50, 500);
+    register Sint count = context->count;
+    register Uint reds = WORK_BEFORE_TRAPPING;
+
+    while (is_list(iter)) {
```
[the remark this answers is not in the document]
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Fixed it, found moments ago too when doing reduction budgets

File: erts/emulator/beam/erl_bif_lists.c [hunk 5]
```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
...
+    Eterm iter_a = s->iter_a; /* load iterator from state */
+    static const Uint WORK_BEFORE_TRAPPING = IF_DEBUG(30, 300);
+    register Uint reds = WORK_BEFORE_TRAPPING;
+    Eterm *b = context->copy_of_b;
+
+    /* A is guaranteed to be proper list, otherwise erts_list_length() call at
+     * start will fail */
+    while (is_list(iter_a)) {
+        const Eterm *ptr_a  = list_val(iter_a);
+        Sint found_at = subtract_find_in_array(CAR(ptr_a), b, context->len_b);
```
[the remark this answers is not in the document]
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: This is potentially a complex state here. Trappable algorithm inside another trappable,. Need to think.

File: erts/emulator/beam/erl_bif_lists.c [hunk 6]
```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
...
+subtract_trappable_scan(Process *p, Eterm A, ErtsSubtractContext *context) {
+    ErtsSubtractScanContext *s = &context->s.scan;
+    Eterm iter_a = s->iter_a; /* load iterator from state */
+    static const Uint WORK_BEFORE_TRAPPING = IF_DEBUG(30, 300);
+    register Uint reds = WORK_BEFORE_TRAPPING;
+    Eterm *b = context->copy_of_b;
+
+    /* A is guaranteed to be proper list, otherwise erts_list_length() call at
+     * start will fail */
+    while (is_list(iter_a)) {
```
[the remark this answers is not in the document]
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Fixed

File: erts/emulator/beam/erl_bif_lists.c [hunk 7]
```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
...
+        case SUBTRACT_STEP_LEN_B: {
+            Eterm res = erts_trappable_list_length(&context->s.len);
+            Sint count;
+            if (res == am_false) {
+                BIF_TRAP2(&subtract_trap_export, p,
+                          SUBTRACT_SPECIAL_VALUE, a_b_state);
+            }
+            count = signed_val(res);;
+            /* Returning true means length is done, result in s.len.count */
+            if (count < 0) {
```
[the remark this answers is not in the document]
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Yes it does. How to return original A, and original B?

File: erts/emulator/beam/erl_bif_lists.c [hunk 8]
```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
...
+        }
+
+        case SUBTRACT_STEP_LEN_B: {
+            Eterm res = erts_trappable_list_length(&context->s.len);
+            Sint count;
+            if (res == am_false) {
+                BIF_TRAP2(&subtract_trap_export, p,
+                          SUBTRACT_SPECIAL_VALUE, a_b_state);
+            }
+            count = signed_val(res);;
```
[the remark this answers is not in the document]
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Fixed

File: erts/emulator/beam/erl_bif_lists.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (garazdawi) [reply erts/emulator/beam/erl_bif_lists.c]: You need to write into the BIF_ARG array.

[the remark this answers is not in the document]
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Fixed

File: erts/emulator/beam/erl_bif_lists.c (hunk 2 above)
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: The problem is I do not know the size of the result beforehand, i.e. i'll have to assume heap is large enough and then allocate after the result is finished. Or if the heap is not large enough I'm in trouble.

File: erts/emulator/beam/erl_bif_lists.c (hunk 1 above)
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Eterm is an int and compiler was ok with this heresy. Thanks, fixed.

File: erts/emulator/beam/erl_bif_lists.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Taking length of B from the budget, instead of `reds--` i am now doing `reds -= len_b` here

**Comment by ferd** [comment]
> Before transforming algorithm into a trappable FSM improved lists:subtract (--) performance by ~11% on random inputs by reversing the algorithm (instead of eliminating elements in A for each element in B, I now shrink B by moving last element in place of eliminated element). On sorted inputs the performance is actually 5-10% lower.

That seems odd. I would have thought that the old algorithm wouldn't have cared about sorted order or not and would be stable. From your description it sounds like your algorithm wouldn't care about sorted order either.

I'm confused as to why sorted input would yield times that are that much worse (~[11% + 5-10%] --> 16-21%?) than random inputs. Either there's something really fishy with the implementation, or the benchmarks are really flawed?

File: erts/emulator/beam/erl_bif_lists.c (hunk 1 above)
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Changing this to 1 and respectively 0 suddenly make algorithm 7 times faster on random inputs of 500 elements each. i think i broke something here.

**Comment by kvakvs** [comment]
@ferd My find in array performs search from the end, if B and A are sorted ascending, then it is worst case scenario most of the times going through all of B to find it.
The benchmark forms 1000 random pairs of arrays of 500..100 elements each so kind of makes this influence less visible.

```
-module(perf).
-export([f/0]).

f() ->
  R = fun rand:uniform/1,
  RR = fun(N) -> [rand:uniform(1000) || _ <- lists:seq(0, N)] end,
  Ranges = [{R(500) + 500, R(500) + 500} || _ <- lists:seq(1, 1000)],
  Times = lists:map(
    fun({A, B}) ->
      {T, _} = timer:tc(
        fun() ->
          lists:subtract(RR(A), RR(B))
        end
      ), T
    end, Ranges),
  Sum = lists:foldl(fun(T, Acc) -> T + Acc end, 0, Times),
  io:format("AVG ~p~n", [Sum / length(Times)]).
```

File: erts/emulator/beam/erl_bif_lists.c (hunk 1 above)
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Yes, a bug in result building. Brb to eat something and fixing

File: erts/emulator/beam/erl_bif_lists.c (hunk 2 above)
Comment (jhogberg) [reply erts/emulator/beam/erl_bif_lists.c]: It's safe to `HAlloc` each element as you go along. This could result in a lot of heap fragments being created, but since you already know the maximum possible length of the result you can use `HAllocX` to reserve (but not allocate) all you need up-front.

File: erts/emulator/beam/external.c (hunk 3 above)
Comment (kvakvs) [reply erts/emulator/beam/external.c]: There's a magical number 3 used, which I copied to my code and it had to become 4 but it didn't and it created some nice segfaults. I gave it a reasonable name now so it won't bite anyone else.

Comment (jhogberg) [reply erts/emulator/beam/external.c]: Both of these changes are good, I just figured they could have their own commit.

File: erts/emulator/beam/erl_bif_lists.c (hunk 2 above)
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Ok let's say I really want to do this and will see if it can be done before I am gone on Sunday (traveling to work at the customer). The idea of not copying and building result in place is a really bright one.

File: erts/emulator/beam/erl_bif_lists.c [hunk 9]
```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
...
+ *
+ * Args: A :: proper_list(), B :: proper_list() - create state and enter stage0
+ * Args: A :: nonvalue(), B :: {list(), list(), magic_binary()} - continue
+ *  The binary contains ErtsSubtractContext struct.
+ *  The step field in state will define next action.
+ */
+static Eterm
+subtract(Process *p, Eterm A, Eterm B) {
+    if (A != SUBTRACT_SPECIAL_VALUE && ! is_tuple(B)) {
+        /* This is initial call entry from Erlang, both args are lists, no ctx
```
[the remark this answers is not in the document]
Comment (kvakvs) [reply erts/emulator/beam/erl_bif_lists.c]: Fixed.
Everything comes at the cost of some more complexity.

**Comment by garazdawi** [comment]
Travis still fails?

**Comment by kvakvs** [comment]
> Travis still fails?

There's a segfault somewhere. I am writing a correctness test and hope to catch it, also the remaining improvement - remove last build result step and build on heap or in a fragment.
